Project init

f-PLT opened the pull request:
Project structure and tools initialization

f-PLT commented:
Not quite done yet, but it's enough to start talking about it!

f-PLT commented:
I could probably keep working on this forever, but I think it's far enough along.

I added specific parts for environment management, as I thinks it's the easiest place to get lost for newcomers and people unfamiliar with either Conda or Poetry.

Let me know what you think, what's lacking and/or hard to comprehend.

liellnima reviewed (changes requested):
Thank you for putting all of this together!!
Things look great, there are just two minor fixes I would like to request (these issues seem only to pop up for conda users), and one thing where I am not sure if it's the expected behavior.

Step-by-step:

- [x] make targets
- [ ] make conda-create-env
- [ ] make conda-poetry-install
- [X] make install
- [X] make check-lint
- [X] make precommit
- [ ] make test

(disclaimer: all the other targets have not been tested by me right now)

**`make conda-create-env`**
In line 114 of .make/base.make a problem arises: The -y tag is not recognized by conda. To run this in conda I did: `conda env create -f environment.yml`

![make_conda-create-env](https://github.com/RolnickLab/ClimateSetExtension/assets/44509330/ec6c0173-2c76-4174-a385-e9ec296aa689)

**`make conda-poetry-install`**:
In line 122 in .make/base.make a problem arises: The poetry package could not be found (PackagesNotFoundError) via the anaconda channels. However, I could install the package with conda-forge instead: `conda install conda-forge::poetry==1.8.2` (works also with conda install -y [..etc..]).

![make_conda-poetry-install](https://github.com/RolnickLab/ClimateSetExtension/assets/44509330/8d2c2a4f-9bd6-43a3-b8c8-ad6762fc7c51)

**`make test`**:
The test fails despite the fact there are no tests to run. See screenshot attached. I am not sure if that error message is expected behaviour since there are no tests yet, or if something else is going on here. 

![make_test](https://github.com/RolnickLab/ClimateSetExtension/assets/44509330/0fc015d7-cd7f-4fe8-8aa8-4dc280f55385)

After those things are fixed, I can approve the PR :)

f-PLT commented:
> Thank you for putting all of this together!! Things look great, there are just two minor fixes I would like to request (these issues seem only to pop up for conda users), and one thing where I am not sure if it's the expected behavior.
> 
> Step-by-step:
> 
> * [x]  make targets
> * [ ]  make conda-create-env
> * [ ]  make conda-poetry-install
> * [x]  make install
> * [x]  make check-lint
> * [x]  make precommit
> * [ ]  make test
> 
> (disclaimer: all the other targets have not been tested by me right now)
> 
> **`make conda-create-env`** In line 114 of .make/base.make a problem arises: The -y tag is not recognized by conda. To run this in conda I did: `conda env create -f environment.yml`
> 
> ![make_conda-create-env](https://private-user-images.githubusercontent.com/44509330/331690039-ec6c0173-2c76-4174-a385-e9ec296aa689.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NzE5NTEsIm5iZiI6MTcxNTk3MTY1MSwicGF0aCI6Ii80NDUwOTMzMC8zMzE2OTAwMzktZWM2YzAxNzMtMmM3Ni00MTc0LWEzODUtZTllYzI5NmFhNjg5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE4NDczMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5NjJmNjZlN2YzYjIwOTNhZmE4ZTEyNDk3MDZiNGJhYjM3NGVmNjVkZjU3N2ZkZjQ3NGI1NTZmNTMzM2IyNDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.L59fEnRPBmH-9sA7CSTKx2HIiKGc-Rpx4_WphViKXdY)
> 
> **`make conda-poetry-install`**: In line 122 in .make/base.make a problem arises: The poetry package could not be found (PackagesNotFoundError) via the anaconda channels. However, I could install the package with conda-forge instead: `conda install conda-forge::poetry==1.8.2` (works also with conda install -y [..etc..]).
> 
> ![make_conda-poetry-install](https://private-user-images.githubusercontent.com/44509330/331690029-8d2c2a4f-9bd6-43a3-b8c8-ad6762fc7c51.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NzE5NTEsIm5iZiI6MTcxNTk3MTY1MSwicGF0aCI6Ii80NDUwOTMzMC8zMzE2OTAwMjktOGQyYzJhNGYtOWJkNi00M2EzLWI4YzgtYWQ2NzYyZmM3YzUxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE4NDczMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1MjlhMmQ1ZTVhMmMxYWIyMTdiNGI3NDc4YjczM2U0YTFjMGQ0OTY1ODVkYjljZmRjNzMyMDVhNjViNDc4ZmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BxhxCwGNlQo1OY_MBpKfQGvhJtO6lJSo3vOIlL8f2Go)
> 
> **`make test`**: The test fails despite the fact there are no tests to run. See screenshot attached. I am not sure if that error message is expected behaviour since there are no tests yet, or if something else is going on here.
> 
> ![make_test](https://private-user-images.githubusercontent.com/44509330/331690045-0fc015d7-cd7f-4fe8-8aa8-4dc280f55385.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU5NzE5NTEsIm5iZiI6MTcxNTk3MTY1MSwicGF0aCI6Ii80NDUwOTMzMC8zMzE2OTAwNDUtMGZjMDE1ZDctY2Q3Zi00ZmU4LThhYTgtNGRjMjgwZjU1Mzg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE3VDE4NDczMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2MjNkOTk4NjAyOTg5YzA1ZTIyOGE2NzY4OTM1MGNiMzlhYWI4YmJiYjgyM2MzY2Y0M2Y2NzEwNWJiN2Q2MzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ngtRznFtIjFBreRoZZ39UzRH33o1fMnvpn0GVs-qjRw)
> 
> After those things are fixed, I can approve the PR :)

Well darn... I was sure I fixed this.. must have crept its way back when I added the possibility of installing to the conda environment without having to activate it.

I should have a fix soon enough

f-PLT commented:
> make conda-create-env In line 114 of .make/base.make a problem arises: The -y tag is not recognized by conda. To run this in conda I did: conda env create -f environment.yml

This is strange, I cannot reproduce this error... (double checked that I did have conda as the CONDA_TOOL)

> make conda-poetry-install: In line 122 in .make/base.make a problem arises: The poetry package could not be found (PackagesNotFoundError) via the anaconda channels. However, I could install the package with conda-forge instead: conda install conda-forge::poetry==1.8.2 (works also with conda install -y [..etc..]).

Hmm ok, I get what happened. I usually consider the setting of defaults channels to include conda-forge, but I get that not necessarily every one would. I'll add it explicitly. This is weird though, because conda-forge is defined as a channel in the `environment.yml` file. I'll push a modification and we'll see if it works on your end.

> make test: The test fails despite the fact there are no tests to run. See screenshot attached. I am not sure if that error message is expected behaviour since there are no tests yet, or if something else is going on here.

This is expected behavior.

liellnima commented:
As you have written me on Slack, the first issue is from a too-old conda version, right? (as we also have it on the cluster)

And for the second thing, I will just wait until you pushed some changes and then I will retry it :) 

And the third thing: great, check.

f-PLT commented:
Alright, ready for some more tests to see if I missed anything.

I did take the opportunity to add for new targets and documentation for `pix` installation, as using it on the cluster requires using a virtual environment, because packages can't be installed with the system's `pip` (no write permissions)

liellnima commented:
Nice! Everything runs smoothly on my local machine!

**Info**
- [X] make targets
- [X] make help
- [X] make version
**Conda targets**
- [X] conda-create-env
- [X] conda-env-info
- [X] conda-poetry-install
- [X] conda-poetry-uninstall
- [X] conda-clean-env
**Poetry targets**
- [X] make poetry-install-auto
- [X] make poetry-install
- [X] make poetry-install-venv
- [X] make poetry-env-info
- [X] make poetry-create-env
- [X] make poetry-remove-env
- [X] make poetry-uninstall-venv
**Install targets**
- [X] make install
- [X] make install-with-lab
**Linting targets**
- [X] make check-lint
- [X] make check-pylint
- [X] make fix-lint
- [X] make precommit
**Test targets**
- [X] make test
- [X] make test-marker
- [X] make test-specific
- [X] make test-custom

I have not tested the versioning targets. The rest seems to work :D 

Would love to get an introduction in the next meeting how to handle the test-marker / test-specific / test-custom :) 

I will approve the merge and we can close this PR.

liellnima reviewed (approved):
See last comment on the thread. The changes have solved all issues, and the PR is approved.

mjfortier reviewed (approved):
LGTM ;)
Appreciate following this back and forth.
